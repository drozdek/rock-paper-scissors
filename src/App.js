import React, { Component } from 'react';
import Player from './Components/Player';
import './App.css';

/**
 * @namespace App
 * @extends React.Component
 * @author [Rafal Drozdowski](https://rdrozdowski.github.io/)
 */
export default class App extends Component {

  /**
   * @constructor 
   * 
   * @param {string} props 
   */
  constructor(props) {
    super(props);
    this.state = {
    };
    this.weapons = ["rock", "paper", "scissors"];
    this.runGame = this.runGame.bind(this);
  }

  /**
   * 
   * @method runGame  starts game
   * 
   * @param none
   */
  runGame() {

    // sets states of weapons used
    this.setState({
      playerWeapon1: this.weapons[Math.floor(Math.random() * 3)],
      playerWeapon2: this.weapons[Math.floor(Math.random() * 3)],
    });

    // fix non immediate mutation
    // of setState method
    setTimeout(() => {
      this.checkWeaponStrength()
    }, 0);
  }

  /**
   * @method checkWeaponStrength  check who has stronger weapon 
   * 
   * @params none
   */
  checkWeaponStrength() {
    const weapon1 = this.state.playerWeapon1,
     weapon2 = this.state.playerWeapon2;

    // in case of weapons have the same values
    if ((weapon1 === weapon2) || (weapon2 === weapon1)) {
      this.setState({
        winner: ' It\'s a tie'
      })
    }

    // in case if paper interacts with scissors or opposite
    else if ((weapon1 === "paper" && weapon2 === "scissors") ||
      (weapon1 === "scissors" && weapon2 === "paper")) {
      this.setState({
        winner: 'Scissors win'
      })
    }

    // in case if paper interacts with rock or opposite
    else if ((weapon1 === "paper" && weapon2 === "rock") ||
      (weapon1 === "rock" && weapon2 === "paper")) {
      this.setState({
        winner: 'Paper wins'
      })
    }

    // in case if scissors interacts with rock or opposite
    else if ((weapon1 === "scissors" && weapon2 === "rock") ||
      (weapon1 === "rock" && weapon2 === "scissors")) {
      this.setState({
        winner: 'Rock wins'
      })
    }
  }

  /**
   * @method render 
   */
  render() {
    return (
      <div className="App">
        {/**
         *instantiate Player components & assign weapon
         */}
        <Player className="App__player App__player--one" playerWeapon={this.state.playerWeapon1} />
        <Player className="App__player App__player--two" playerWeapon={this.state.playerWeapon2} />

        {/**
         * display result
         */}
        <p><b>{typeof this.state.winner === "undefined" ? 'Have a play, please...' : this.state.winner}</b></p>
        
        {/**
        *show clickable button
        */}
        <button className="App__button" onClick={this.runGame}> Play </button>
      </div>
    );
  }

}

