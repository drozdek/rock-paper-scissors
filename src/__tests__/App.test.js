import React from 'react';
import { mount, shallow } from 'enzyme';
import ReactDOM from 'react-dom';
import App from '../App';
import { create } from 'react-test-renderer'

describe('Rock, paper, scissors game', () => {
  test('creates snapshot', () => {
    let tree = create(<App />)
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has Play button', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.text().includes('Play')).toBe(true);
  });

  test('simpulates click', () => {
    const wrapper = shallow(<App />);
    wrapper.find('button').simulate('click');
  })
});

