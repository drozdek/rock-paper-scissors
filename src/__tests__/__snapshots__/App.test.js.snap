// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rock, paper, scissors game creates snapshot 1`] = `
<div
  className="App"
>
  <div
    className="App__player"
  >
    <div>
      <img
        alt="Image of undefined.png"
        className="App__image--adjust"
        src="undefined.png"
      />
    </div>
  </div>
  <div
    className="App__player"
  >
    <div>
      <img
        alt="Image of undefined.png"
        className="App__image--adjust"
        src="undefined.png"
      />
    </div>
  </div>
  <p>
    <b>
      Have a play, please...
    </b>
  </p>
  <button
    className="App__button"
    onClick={[Function]}
  >
     Play 
  </button>
</div>
`;
